accounts-db/src/accounts.rs

```diff
@@ -552,31 +569,52 @@ impl Accounts {
         account_locks: &mut AccountLocks,
         writable_keys: Vec<&Pubkey>,
         readonly_keys: Vec<&Pubkey>,
-    ) -> Result<()> {
+        entry_accts_lookup: &mut EntryAcctLocks,
+        allow_self_conflicting_entries: bool,
+    ) -> (Result<()>, bool) {
+        let mut self_conflicting_account: bool = false;
+        let mut self_conflicting_tx = false;
         for k in writable_keys.iter() {
             if account_locks.is_locked_write(k) || account_locks.is_locked_readonly(k) {
-                debug!("Writable account in use: {:?}", k);
-                return Err(TransactionError::AccountInUse);
+                if allow_self_conflicting_entries {
+                    self_conflicting_account = entry_accts_lookup.writables.contains(k)
+                        || entry_accts_lookup.readables.contains(k);
+                }
+                if !self_conflicting_account {
+                    debug!("Writable account in use: {:?}", k);
+                    return (Err(TransactionError::AccountInUse), self_conflicting_tx);
+                } else {
+                    self_conflicting_tx = true;
+                }
             }
         }
         for k in readonly_keys.iter() {
             if account_locks.is_locked_write(k) {
-                debug!("Read-only account in use: {:?}", k);
-                return Err(TransactionError::AccountInUse);
+                if allow_self_conflicting_entries {
+                    self_conflicting_account = entry_accts_lookup.writables.contains(k);
+                }
+                if !self_conflicting_account {
+                    debug!("Read-only account in use: {:?}", k);
+                    return (Err(TransactionError::AccountInUse), self_conflicting_tx);
+                } else {
+                    self_conflicting_tx = true;
+                }
             }
         }
 
         for k in writable_keys {
             account_locks.write_locks.insert(*k);
+            entry_accts_lookup.writables.insert(*k);
         }
 
         for k in readonly_keys {
             if !account_locks.lock_readonly(k) {
                 account_locks.insert_new_readonly(k);
+                entry_accts_lookup.readables.insert(*k);
```

Comment:
i think this should be put outside `if !account_locks.lock_readonly(k)`?

iirc, `account_locks` state here is global? i.e. given account can be read-only already by an unrelated in-flight batch.

Reply:
yeah. though interface on account locks is bad here imo.
why is this not just `insert_read_lock(k)`, there's no reason the calling code (here) needs to know if it's a new entry vs old entry.

Reply:
sorry if comment was not clear. @Huzaifa696 , I'm not asking you to change that interface in this PR, was just pointing out the interface (imo) is confusing.

Just make the change @ryoqun suggested and that will be good here.

Reply:
right, I am on it. other than that I am just writing `insert_read_lock` and `insert_write_lock` functions to modify `BatchAccountLocks`